Copy missing tool headers and fix script/ binaries

ljskernel opened the pull request:
Fixes #3.
## Add missing tool headers

`arch/<arch>/tools` is a small directory that the `scripts` build requires and additionally might be required by kernel module builds. `tools/include/...` is definitely required and has actively broken a user's kernel build so we include this also (but not all of `tools/...`  as that code is 15MiB in size.)
## Add workaround to generate script/ binaries in target arch

If you compile without any special provisions binaries in this directory will be host arch binaries.

If you attempt to force matters by setting HOSTCC, you face the issue that certain of the generated binaries need to be run to generate other binaries, and a catch-22 emerges.

The solution implemented here is to build as normal, then move the script binaries to the 'xscripts' directory. When the module build environment is established, Makefile's that define what scripts are run are backed up then adjusted to run binaries from xscripts, then the scripts directory is built to generate binaries for the target architecture. Once this is complete the Makefiles are made pristine again and xscripts is removed.

ljskernel committed:
Copy over arch-specific tools and tools headers.

Some kernel modules appear to require these esp. on ARM. Don't copy _entire_
tools/ directory as that is 15MiB.

ljskernel committed:
Create helper push function.

ljskernel committed:
Add workaround to generate script/ binaries in target arch.

If you compile without any special provisions binaries in this directory will be
host arch binaries.

If you attempt to force matters by setting HOSTCC, you face the issue that
certain of the generated binaries need to be run to generate other binaries, and
a catch-22 emerges.

The solution implemented here is to build as normal, then move the script
binaries to the 'xscripts' directory. When the module build environment is
established, Makefile's that define what scripts are run are backed up then
adjusted to run binaries from xscripts, then the scripts directory is built to
generate binaries for the target architecture. Once this is complete the
Makefiles are made pristine again and xscripts is removed.

lorenzo-stoakes commented:
Have tested against kernel source for rpi3, header generation works and additionally module build on target device works.

lorenzo-stoakes commented:
Tried build against apalis-imx6 confirmed working.

lorenzo-stoakes commented:
Tried header generation against colibri-imx6, confirmed working.

ljskernel committed:
Only do scripts bin dance if we are cross-compiling.

lorenzo-stoakes commented:
Tested against x86 w/new change to not attempt the script bins dance in this case, confirmed working.

ljskernel committed:
Make objects dir a required argument.

Currently the script is broken if object dirs is not specified, since it will
attempt to use O= with a non-pristine tree and if you `make mrproper` the tree
.config and Module.symvers disappear.

It's simpler for now to simply make this a required parameter, perhaps later I
can adjust the script to run without this separation.

ljskernel committed:
Update README.

lorenzo-stoakes commented:
Fixes #7

agherzan commented on gen_mod_headers:
```diff
@@ -140,4 +145,41 @@ for f in $(find . -name "Kconfig*"); do
+push "$target_dir"
+
+# Make some backups so we can restore sanity after we're done.
+cp scripts/basic/Makefile .backup.basic.Makefile
```
You copy scripts in xscripts and then you backup some files. Why exactly? Cause you will have them in xscripts if needed.

ljskernel replied:
@agherzan It doesn't actually copy in anything to `xscripts`, xscripts receives ONLY the binaries for the host arch because `make O=...` is invoked, we copy in the source to `scripts` later.

agherzan replied:
Got it.

agherzan reviewed (changes requested):
Looks good
